File: scripts/agent_launcher.py
#!/usr/bin/env python3
import argparse
import os
import subprocess
import sys

DEFAULT_GOPASS_PREFIX = "mentci/ai"
DEFAULT_COMMAND = ["opencode"]
ENV_BY_PROVIDER = {
    "deepseek": "DEEPSEEK_API_KEY",
    "openai": "OPENAI_API_KEY",
    "anthropic": "ANTHROPIC_API_KEY",
}


def _read_gopass_secret(entry: str) -> str:
    try:
        result = subprocess.run(
            ["gopass", "show", "--password", entry],
            check=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )
    except FileNotFoundError:
        raise RuntimeError("gopass not found in PATH") from None
    except subprocess.CalledProcessError as exc:
        detail = exc.stderr.strip() or exc.stdout.strip()
        raise RuntimeError(f"gopass failed for {entry}: {detail}") from None
    return result.stdout.strip()


def _resolve_entry(prefix: str, provider: str, entry_override: str | None) -> str:
    if entry_override:
        return entry_override
    return f"{prefix}/{provider}/api-key"


def _resolve_env_var(provider: str, env_override: str | None) -> str:
    if env_override:
        return env_override
    env_var = ENV_BY_PROVIDER.get(provider)
    if not env_var:
        raise RuntimeError(
            f"Unknown provider '{provider}'. Use --env-var or extend ENV_BY_PROVIDER."
        )
    return env_var


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Launch an AI agent with API keys pulled from gopass."
    )
    parser.add_argument("--provider", required=True, help="Provider name (deepseek).")
    parser.add_argument(
        "--gopass-prefix",
        default=os.environ.get("MENTCI_GOPASS_PREFIX", DEFAULT_GOPASS_PREFIX),
        help="Entry prefix for gopass (default: mentci/ai).",
    )
    parser.add_argument(
        "--entry",
        help="Explicit gopass entry override (default: <prefix>/<provider>/api-key).",
    )
    parser.add_argument(
        "--env-var",
        help="Explicit environment variable name for the API key.",
    )
    parser.add_argument(
        "command",
        nargs=argparse.REMAINDER,
        help="Command to execute (default: opencode).",
    )
    args = parser.parse_args()

    entry = _resolve_entry(args.gopass_prefix, args.provider, args.entry)
    env_var = _resolve_env_var(args.provider, args.env_var)
    api_key = _read_gopass_secret(entry)

    cmd = args.command if args.command else DEFAULT_COMMAND
    if cmd and cmd[0] == "--":
        cmd = cmd[1:]
    if not cmd:
        cmd = DEFAULT_COMMAND

    env = os.environ.copy()
    env[env_var] = api_key
    return subprocess.call(cmd, env=env)


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/agent_launcher.clj
#!/usr/bin/env bb

(require '[babashka.deps :as deps])
(deps/add-deps '{:deps {metosin/malli {:mvn/version "0.17.0"}}})

(ns agent-launcher
  (:require [clojure.string :as str]
            [clojure.java.shell :as sh]
            [malli.core :as m]
            [malli.error :as me]
            [clojure.java.io :as io]))

(load-file (str (.getParent (io/file *file*)) "/types.clj"))

(def default-gopass-prefix "mentci/ai")
(def default-command ["opencode"])
(def env-by-provider
  {"deepseek" "DEEPSEEK_API_KEY"
   "openai" "OPENAI_API_KEY"
   "anthropic" "ANTHROPIC_API_KEY"})

(defn fail [msg]
  (binding [*out* *err*]
    (println msg))
  (System/exit 1))

(defn parse-args [args]
  (loop [remaining args
         {:keys [provider] :as opts} {}]
    (if (empty? remaining)
      opts
      (let [[arg & more] remaining]
        (cond
          (= arg "--provider")
          (recur (nnext remaining) (assoc opts :provider (second remaining)))

          (= arg "--gopass-prefix")
          (recur (nnext remaining) (assoc opts :gopass-prefix (second remaining)))

          (= arg "--entry")
          (recur (nnext remaining) (assoc opts :entry (second remaining)))

          (= arg "--env-var")
          (recur (nnext remaining) (assoc opts :env-var (second remaining)))

          (= arg "--")
          (assoc opts :command (vec more))

          :else
          (assoc opts :command (vec remaining)))))))

(defn resolve-entry [prefix provider entry]
  (or entry (str prefix "/" provider "/api-key")))

(defn resolve-env-var [provider env-var]
  (or env-var
      (get env-by-provider provider)
      (fail (str "Unknown provider '" provider "'. Use --env-var or extend env-by-provider."))))

(defn gopass-secret [entry]
  (let [{:keys [exit out err]} (sh/sh "gopass" "show" "--password" entry)]
    (when-not (zero? exit)
      (fail (str "gopass failed for " entry ": " (str/trim (or err out)))))
    (str/trim out)))

(defn validate-config [config]
  (when-not (m/validate types/AgentLauncherConfig config)
    (throw (ex-info "Invalid Agent Launcher Configuration"
                    {:errors (me/humanize (m/explain types/AgentLauncherConfig config))}))))

(defn -main [& args]
  (let [{:keys [provider gopass-prefix entry env-var command]}
        (parse-args args)
        provider (or provider (fail "Missing --provider"))
        prefix (or gopass-prefix (System/getenv "MENTCI_GOPASS_PREFIX") default-gopass-prefix)
        entry (resolve-entry prefix provider entry)
        env-var (resolve-env-var provider env-var)
        cmd (if (seq command) command default-command)
        cmd (if (= (first cmd) "--") (vec (rest cmd)) cmd)
        cmd (if (seq cmd) cmd default-command)
        config {:sema/type "AgentLauncherConfig"
                :provider provider
                :gopassPrefix prefix
                :entry entry
                :envVar env-var
                :command cmd}
        _ (validate-config config)
        api-key (gopass-secret entry)]
    (let [env (assoc (into {} (System/getenv)) env-var api-key)
          {:keys [exit]} (apply sh/sh {:env env} cmd)]
      (System/exit exit))))

(apply -main *command-line-args*)
